# <center>Laboratorium 1<br>Metody Numeryczne</center>

Biblioteki niezbędne do wykonania zadania:

(wykonanie tego bloku zapewnia do nich dostęp w kolejnych blokach)

In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-
import main
import numpy as np
import scipy
import pandas as pd
import math

***Zadanie 1.***  
Zaimplementuj funkcję $cylinder\_area$ tak by zwracała pole powierzchni walca o promieniu podstawy $r$ i wysokości $h$. Stała $\pi$ jest zdefiniowana np. w bibliotece *[math](https://docs.python.org/3/library/math.html#constants)*. Jeżeli nie da się policzyć pola funkcja powinna zwracać wartość *NaN*. Sprawdź działanie zaimplementowanej funkcji dla dowolnych wartości w tym notatniku.

In [2]:
print(main.cylinder_area(10,10))
print(main.cylinder_area(-10,10))

1256.6370614359173
nan


***Zadanie 2.***  
Wygeneruj dowolne ciągi arytmetyczny o kroku różnym od jeden i niebędącym liczbą całkowitą, używając w tym celu funkcji *[arange](https://docs.scipy.org/doc/numpy/reference/generated/numpy.arange.html)* oraz *[linspace](https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.linspace.html)*. Wypisz go i przeanalizuj otrzymane rezultaty.

In [3]:
#Za pomoca fukcji arange
print("Za pomoca funkcji arange:\n",np.arange(0.091, 1.0, 0.1))
#Za pomoca funkcji linspace
print("Za pomoca funkcji linspace:\n",np.linspace(0.091, 1.0, num = 6, endpoint=False))

Za pomoca funkcji arange:
 [0.091 0.191 0.291 0.391 0.491 0.591 0.691 0.791 0.891 0.991]
Za pomoca funkcji linspace:
 [0.091  0.2425 0.394  0.5455 0.697  0.8485]


***Komentarz***  
Uzycie funkcji arange umozliwia tworzenia tablicy z zadaniem poczatku, konca oraz kroku.
W prosty sposob mozna tworzyc "kontrolowane" przez nas listy

Troszke inne rezultaty daje funkcja linspace, gdzie zadalem argumenty poczatku, konca oraz ilosci elementow.
Za pomoca tej funkcji mozna podzielic tablice po rowno nie muszac przeliczac wczesniej

***Zadanie 3.***  
Zaimplementuj funkcję $fib$ zwracającą wektor pierszych n elementów *[ciągu Fibonnaciego](https://pl.wikipedia.org/wiki/Ci%C4%85g_Fibonacciego)*, jeżeli nie jest to możliwe funkcja powinna zwrócić wartość *None*.

In [5]:
print(main.fib(7))   #7 pierwszych wyrazow
print(main.fib(-2))
print(main.fib(3.0))

[[ 1.  1.  2.  3.  5.  8. 13.]]
None
None


***Komentarz***  
Za pomocą komendy print wypisałem 7 pierwszych wyrazów ciągu fibonnaciego z celowo powtórzoną jedynką. Za pomocą odkrytej funkcji np.append dodawałem co każde przejście pętli jeden wyraz ciągu fibonacciego do wektora tworząc całość 

***Zadanie 4.***  
Napisz funkję $matrix\_calculations$ która jako argument przyjmuje dowolną wartość liczbową *a* i tworzy macierz:  
  
$\mathbf{M} =\left[ \begin{matrix}
a & 1 & -a\\
0 & 1 & 1\\
-a & a & 1
\end{matrix}\right]
$    
   
Dla zadeklarowanej macierzy wyznacz numerycznie macierz odwrotną $\mathbf{Minv}$ (jeżeli nie istnieje taka macierz wartość wynosi NaN), macierz transponowaną $\mathbf{Mt}$ i wyznacznik macierzy $\mathbf{Mdet}$. Zwróć otrzymane wartości w postaci krotki postaci $(\mathbf{Minv}, \mathbf{Mt}, \mathbf{Mdet})$.  Wypisz otrzymane wyniki. 
  
***Wskazówki:*** Do tworzenia obiektów mających własności macierzy w języku Python używa się klasy *[array](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.array.html)* z pakietu *[numpy](http://www.numpy.org/)*, przykładowe użycie:

In [6]:
main.matrix_calculations(5)

(array([[ 0.08,  0.52, -0.12],
        [ 0.1 ,  0.4 ,  0.1 ],
        [-0.1 ,  0.6 , -0.1 ]]),
 array([[ 5,  0, -5],
        [ 1,  1,  5],
        [-5,  1,  1]]),
 -49.99999999999997)

***Komentarz***  
Miejce, które może zwrócić uwage to wyliczony wyznacznik macierzy jako trzeci element krotki, który normalnie wynosi 50, a za pomocą funkcji np.linalg.det() jest jedynie bliski tej liczby, prawdopodobnie jest to przybliżenie ale nie mam do końca pewności dlaczego tak wychodzi

***Zadanie 5.***  
Stwórz w noataniku macierz:  
$\mathbf{M} =\left[ \begin{matrix}
3 & 1 & -2 & 4\\
0 & 1 & 1 & 5\\
-2 & 1 & 1 & 6\\
4 & 3 & 0 & 1\\
\end{matrix}\right]
$   
Wypisz przy pomocy funkcji *print* następujące elementy macierzy $\mathbf{M}$: $\mathbf{M}_{1,1},\mathbf{M}_{3,3},\mathbf{M}_{3,2}$. Zdefiniuj wektor $\mathbf{w1}$ którego elementy to trzecia kolumna macierzy $\mathbf{M}$ oraz wektor $\mathbf{w2}$ który składa się z drugiego wiersza tej macierzy.  
  
***Wskazówki:***
Tablice z pakietu *numpy* są indeksowane od zera a do każdego elementu można odwołać się poprzez jego indeks. Przykład użycie


In [7]:
m = np.array([[3,1,-2,4],[0,1,1,5],[-2,1,1,6],[4,3,0,1]])
print(m[1,1])
print(m[3,3])
print(m[3,2])
w1 = m[:,2]
w2 = m[1,:]

1
1
0


Dodatkowo twórcy biblioteki umożliwiają użytkownikowi na manipulację elementami tablicy poprzez operator ':', więcej szczegółów na temat jego użycia w artykule [NumPy for Matlab users](https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html#numpy-for-matlab-users).

***Zadanie 6.***   
Uzupełnij funkcję $custom\_matrix$, tak by parametry m, n określały wymiary macierzy wynikowej, która
będzie wypełniona w/g algorytmu: jeśli indeks wiersza jest większy od indeksu kolumny
wartością komórki jest indeks wiersza, w przeciwnym wypadku wartością komórki jest indeks
kolumny. Funkcja jako wynik powinna zwracać uzupełnioną macierz, jeżeli nie jest to możliwe to powinna zwrócić *None*. Na koniec wyświetlić wynikową macierz dla dowolnych argumentów $m$, $n$ z przedziału $\langle3,7\rangle$.
  
***Wskazówka:*** Inicjalizacja pustej macierz wykonywana jest w pakiecie Numpy przy pomocy funkcji  *[zeros](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.zeros.html)*, zaś macierzy składającej się z jedynek *[ones](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ones.html)*.

In [10]:
print(main.custom_matrix(4,6))
print(main.custom_matrix(4.0,6))

[[0. 1. 2. 3. 4. 5.]
 [1. 1. 2. 3. 4. 5.]
 [2. 2. 2. 3. 4. 5.]
 [3. 3. 3. 3. 4. 5.]]
None


***Komentarz***  
Za pomocą funkcji np.zeros() zainicjalizowałem macierz o wymiarach m x n wypełniając ją zerami, a do funkcji przekazałem krotke. Nastepnie dla każego indeksu sprawdzałem warunki i zmienialem wartości 

***Zadanie 7.***

Biblioteka Numpy posiada własne metody, pozwalające na szybsze i wygotniejsze wykonywanie operacji na wektorach i macierzach. Kolejne zadania mają na celu przećwiczenie tych metod. Zamiast samemu implementować mnożenie macierzy, należy skorzystać właśnie z wbudowanych funkcji Numpy. Najbardzie podstawowe z nich to *[np.multiply](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.multiply.html)*, *[np.dot](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.dot.html)* oraz *[np.matmul](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.matmul.html)*. Przed wykonaniem zadania należy zapoznać się z ich dokumentacją, aby stosować poprawną funkcję do danego typu mnożenia. Dodatkowo ciekawą i użyteczną funkcjonalnością Numpy, wykorzystywaną niekiedy przy dodawaniu macierzy jest *broadcasting*, proszę o zapoznaie się z opisem:  *[[1]](https://docs.scipy.org/doc/numpy-1.15.0/user/basics.broadcasting.html)* *[[2]](https://www.tutorialspoint.com/numpy/numpy_broadcasting.htm)*.

Zainicjalizować dwa wektory $v_1$ i $v_2$:

$v_1 = \begin{bmatrix}1\\3\\13\end{bmatrix} \quad v_2 = \begin{bmatrix}8\\5\\-2\end{bmatrix}$

Następnie wykonać operacje i wypisać ich wynik:
- $4*v_1$
- $-v_2 + \begin{bmatrix}2\\2\\2\end{bmatrix}$
- $v_1 \circ v_2$ (w sensie mnożenia macierzy, tzw. mnożenie macierzy w sensie Cauchy’ego)
- $v_1 * v_2$ (w sensie mnożenia Hadamarda (element-wise))

***Wskazówki:*** Warto wiedzieć o *[np.dot](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.dot.html)* i *[np.multiply](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.multiply.html)*. Zbadać jak zachowuje się przeciążony operator mnożenia $*$ oraz $@$ dla macierzy $ndarray$.

In [8]:
v1 = np.array([[1],[3],[13]])
v2 = np.array([[8],[5],[-2]])        
#pierwsza operacja
operation_1 = v1*4                                 #skorzystalem z broadcastingu, bo wymiary sie zgadzaja
print(operation_1,"\n")
#druga operacja
operation_2 = -v2 + np.array([[2],[2],[2]])         
print(operation_2,"\n")
#trzecia operacja
operation_3 = np.matmul(np.transpose(v1), v2)
print(operation_3,"\n")

operation_4 = v1*v2
print(operation_4,"\n")
#Przeciazenie operatora *
print("Przeciazenie operatora na przykladzie wektorow v1 i v2*\n", v1*v2)
#Stworze nowy wektor aby sprawdzic przeciazenie operatora @
v3 = np.array([1, 2, 3])
print("Przeciazenie operatora @ na przykladzie wektorow v1 i v3\n",v3@v1)

[[ 4]
 [12]
 [52]] 

[[-6]
 [-3]
 [ 4]] 

[[-3]] 

[[  8]
 [ 15]
 [-26]] 

Przeciazenie operatora na przykladzie wektorow v1 i v2*
 [[  8]
 [ 15]
 [-26]]
Przeciazenie operatora @ na przykladzie wektorow v1 i v3
 [46]


***Komentarz***  
W tym zadaniu nie bylo konieczne skorzystanie z funkcji multiply, wystarczylo korzystac z operatorow * oraz @, ale dla przykladu skorzystalem z funkcji np.matmul(), ale mnozenie w sensie Cauchyego nie bylo mozliwe z powodu roznych wymiarow macierzy, dlatego musiałem odwrócić jeden z wektorów  
Dzieki przeciazaniu operatorow moglem w prosty sposob wykonac dzialania na macierzach

***Zadanie 8.***
Zainicjalizować macierz $M_1$:

$M_1 = \begin{bmatrix}1&-7&3\\-12&3&4\\5&13&-3\end{bmatrix}$

Następnie wykonać operacje i wypisać ich wynik:
- $3M_1$
- $3M_1 + \begin{bmatrix}1&1&1\\1&1&1\\1&1&1\end{bmatrix}$
- $M_1^T$
- $M_1 \circ v_1$ 
- $v_2^T \circ M_1$

In [10]:
M1 = np.array([[1,-7,3],[-12,3,4],[5,13,-3]])
print("Pierwsza operacja")
print(M1*3)
print("\nDruga operacja")
print(M1*3 + np.ones((3,3)))
print("\nTrzecia operacja")
print(M1.transpose())
print("\nCzwarta operacja")
v1 = np.array([[1],[3],[13]])
v2 = np.array([[8],[5],[-2]])
print(M1@v1)
print("\nOstatnia operacja")
print(np.transpose(v2)@M1)

Pierwsza operacja
[[  3 -21   9]
 [-36   9  12]
 [ 15  39  -9]]

Druga operacja
[[  4. -20.  10.]
 [-35.  10.  13.]
 [ 16.  40.  -8.]]

Trzecia operacja
[[  1 -12   5]
 [ -7   3  13]
 [  3   4  -3]]

Czwarta operacja
[[19]
 [49]
 [ 5]]

Ostatnia operacja
[[-62 -67  50]]


***Komentarz***  
Wszystkie operacje były możliwe do wykonania za pomocą prostych funkcji i operatorów, warto też zauważyć w jaki sposób został wypisany ostatni wektor, w podwójnych nawiasach.

# Wnioski
-Dane ćwiczenia są dobre na stopniowe wprowadzenie do biblioteki numpy i pandas i zaznajomienie z interpreterem Anaconda, bo na takim pracowałem w Pycharmie.   
-Zapoznałem sie z notatnikiem Jupyter, gdzie można w sposób czytelny przedstawiać i komentować kod programistyczny  
-Biblioteka numpy ułatwia tworzenie wektorów i macierzy, gdyż nie musimy tworzyć petli w petli i różnych kombinacji  
-Problematyczne mogą okazać się wartości jakie przekazujemy do różnych funkcji, takie jak np. przy mnożeniu macierzy w sensie Cauchyego musimy pamiętać o tych samych wymiarach, lub w funkcjach arange() i linspace() żeby nie mylić liczb, ale odpowiedni błąd wyskakuje co można łatwo nadrobić.  
-Poznałem wartość numeryczną taką jak NaN - Not a number -która służy do oznaczania wartości niezdefiniowanej lub niereprezentowalej, zwłaszcza w obliczeniach wykorzystujących liczby zmiennoprzecinkowe.  
-Przez przypadek odkryłem, aby odwrócić wektor poziomy za pomocą funkcji np.transpose() musze do jej wnętrza dodać nawias kwadratowy, na zasadzie np.transpose( [np.array([1,2,3])] ), lub do stworzonego poziomego wektora dodać ten nawias  
-Warto też zwrócić uwage na obliczanie wyznacznika macierzy za pomocą funkcji np.linalg.det(), ponieważ w zadaniu 4 licząc kalkulatorem czy ręcznie wynosi równe -50, a funkcja wyżej zwraca -49.99999999999997, czyli nieprzybliżoną wartość
-Do stworzonych macierzy za pomocą biblioteki numpy można w łatwy sposób odwoływać się do odpowiednich indeksów i manipulować nimi, tak jak to jest w tablicach stworzonych za pomocą list  
-W mainie aby każda funkcja działała poprawnie, tj. zwracała None jak w poleceniach przy ewentualnych błędach wpisania, np. w funkcji fib() musiałem pamietać, że wpisanie floata powinno zwrócić None, aby problem załatwić poznałem nową funkcje isinstance(n,int) która sprawdza czy n jest int

Materiały uzupełniające:
- [Scipy Lecture Notes](http://www.scipy-lectures.org/index.html)
- [NumPy for Matlab users](https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html#numpy-for-matlab-users)
- [Python Tutorial - W3Schools](https://www.w3schools.com/python/default.asp)
- [NumPy](https://www.numpy.org)
- [Matplotlib](https://matplotlib.org/)
- [Anaconda](https://www.anaconda.com/)
- [Learn Python for Data Science](https://www.datacamp.com/learn-python-with-anaconda?utm_source=Anaconda_download&utm_campaign=datacamp_training&utm_medium=banner)
- [Learn Python](https://www.learnpython.org/)
- [Wujek Google](https://google.pl) i [Ciocia Wikipedia](https://pl.wikipedia.org/wiki/Wikipedia:Strona_g%C5%82%C3%B3wna)